components/modules/rmt.c

```diff
@@ -0,0 +1,403 @@
+  config.clk_div = divisor;
+
+  if (lua_type(L, 3) == LUA_TTABLE) {
+    lua_getfield(L, 3, "carrier_hz");
+    int hz = lua_tointeger(L, -1);
+    if (hz) {
+      config.tx_config.carrier_freq_hz = hz;
+      config.tx_config.carrier_en = true;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "carrier_duty");
+    int duty = lua_tointeger(L, -1);
+    if (duty) {
+      config.tx_config.carrier_duty_percent = duty;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "idle_level");
+    if (!lua_isnil(L, -1)) {
+      int level = lua_tointeger(L, -1);
+      config.tx_config.idle_level = level;
+      config.tx_config.idle_output_en = true;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "invert");
+    if (lua_toboolean(L, -1)) {
+      config.flags |= RMT_CHANNEL_FLAGS_INVERT_SIG;
+    }
+    lua_pop(L, 1);
+  }
+
+  lrmt_channel_t ud = (lrmt_channel_t)lua_newuserdata(L, sizeof(*ud));
+  if (!ud) return luaL_error(L, "not enough memory");
+  memset(ud, 0, sizeof(*ud));
+  luaL_getmetatable(L, "rmt.channel");
+  lua_setmetatable(L, -2);
+
+  // We have allocated the channel -- must free it if the rest of this method fails
+  int channel = platform_rmt_allocate(1, RMT_MODE_TX);
+
+  if (channel < 0) {
+    return luaL_error(L, "no spare RMT channel");
+  }
+
+  config.channel = channel;
+
+  ud->channel = channel;
+  ud->tx = true;
+
+  int rc = rmt_config(&config);
+  if (rc) {
+    platform_rmt_release(config.channel);
+    return luaL_error(L, "Failed to configure RMT");
+  }
+
+  rc = rmt_driver_install(config.channel, 0, 0);
+  if (rc) {
+    platform_rmt_release(config.channel);
+    return luaL_error(L, "Failed to install RMT driver");
+  }
+
+  lua_pushinteger(L, divisor * 12500);
+
+  return 2;
+}
+
+static int lrmt_rxsetup(lua_State *L) {
+  int gpio = luaL_checkinteger(L, 1);
+  int divisor = get_divisor(L, 2);
+
+  // We will set the channel later
+  rmt_config_t config = RMT_DEFAULT_CONFIG_RX(gpio, 0);
+  config.clk_div = divisor;
+  config.rx_config.idle_threshold = 65535;
+
+  if (lua_type(L, 3) == LUA_TTABLE) {
+    lua_getfield(L, 3, "invert");
+    if (lua_toboolean(L, -1)) {
+      config.flags |= RMT_CHANNEL_FLAGS_INVERT_SIG;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "filter_ticks");
+    if (!lua_isnil(L, -1)) {
+      int ticks = lua_tointeger(L, -1);
+      if (ticks < 0 || ticks > 255) {
+        return luaL_error(L, "filter_ticks must be in the range 0 - 255");
+      }
+      config.rx_config.filter_ticks_thresh = ticks;
+      config.rx_config.filter_en = true;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "idle_threshold");
+    if (!lua_isnil(L, -1)) {
+      int threshold = lua_tointeger(L, -1);
+      if (threshold < 0 || threshold > 65535) {
+        return luaL_error(L, "idle_threshold must be in the range 0 - 65535");
+      }
+      config.rx_config.idle_threshold = threshold;
+    }
+    lua_pop(L, 1);
+  }
+
+  lrmt_channel_t ud = (lrmt_channel_t)lua_newuserdata(L, sizeof(*ud));
```

Comment:
Aside from RX/TX mode and RX buffer size, the second half of this function looks identical to the TX setup - may I suggest factoring it out into a single function to save some space?